### ISSUE TITLE
Question: comparison to other tools

### ISSUE THREAD

**Issue by lonix1** [jameswlane/jest-express#176]
New to node (and thus to testing with jest, etc.)

I need to test my express app, using mocks and so on. I've seen lots of projects use "supertest". And jest and supertest can work together.

Can someone please tell me what is the difference between this and supertest? Do they do the same thing?

**Comment by jameswlane** [comment]
@lonix1 

You would use `jest-express` for doing TDD/BDD type testing or unit testing. That would allow you to test your code faster in the process and in smaller chunks.

You would use something like `supertest` for integration testing. Generally, you would have a smaller number of integration test due to they only test the endpoints of your application vs the entire application. They are also slower due to they have to spin up your application and express and when the test are complete you have to tear the application down.

When I am personally building a express application I have multiple layers of testing. Unit (testing the code), Integration (testing the API's), Acceptance (testing UI layer with no backend calls), Smoke (testing the API when deployed on a server), and End to End (testing the entire application when deployed out on a server). That generally gives you the best coverage over an application, everything from end cases to happy path workflows.

**Comment by lonix1** [comment]
I'm new to node but not to BDD/TDD/etc., and so much of what's written out there is annoying and not specific - e.g. supertest's description is "HTTP assertions made easy via superagent." Very helpful :)

Your analysis was spot-on, that is how I test too. I spent a few hours playing with supertest and it seems that all it does is simplify http requests - but I need to setup/teardown the server for each test, so from what I can tell it's not mocking the requests. It's just an easy to use request library. Am I correct in that analysis?

Whereas your library is described as "Mock Express for testing with Jest" - I only saw that now... would've saved lots of time if I'd paid more attention. :)

So in summary, for my benefit and future readers, please tell me if you agree:
- `jest-express` mocks Express, so: 1) no setup/teardown, 2) I can remain ignorant of express and http.Server, and 3) I can unit tests my endpoints in isolation
- `supertest` doesn't mock, it's a *convenience library* that simply wraps Express and http.Server (thus simplifying integration tests / e2e tests). It's important to remember there is still a real server underneath.

**Comment by jameswlane** [comment]
@lonix1,
`jest-express` is not made for testing endpoints and it does no HTTP calls.
It allows you to mock out all of express and test your application's code without the need of starting up an express server. 

For example: [Here](https://github.com/jameswlane/status-board/blob/master/test/unit/webapp/routes/widget/log.test.ts) I am testing to make sure the render function uses the correct template for logging.

I am testing internal methods that wouldn't be exposed to http calls.

But the bullet point for `supertest` is spot on.

**Comment by lonix1** [comment]
Yes that's what I meant - to test the endpoints in isolation without caring about http.

Which is exactly what I need, so many thanks for this library!

**Comment by lonix1** [comment]
Sorry last question, since you're an expert in these matters, and it's related to your library. Are there any other recommended test tools? (I mean for the "big stuff", not little convenience tools of which there seem to be 2 bazillion on npm.)

I'm using Jest (obviously), jest-express (unit) and supertest (integration).

Is anything else typically used that a newbie should look into - "protractor" for example?
